datafusion/core/src/physical_plan/sorts/cascade.rs

```diff
@@ -0,0 +1,305 @@
+/// │  2  │                │  4  │                │                                          
+/// └─────┘                └─────┘                                                           
+/// ┌─────┐                ┌─────┐                ▼                                          
+/// │  1  │                │  1  │                                                           
+/// │  4  │─ ▶  sort  ─ ─ ▶│  1  ├ ─ ─ ─ ─ ─ ▶ merge  ─ ─ ─ ─                                
+/// │  1  │                │  4  │                           │                               
+/// └─────┘                └─────┘                                                           
+///   ...                   ...                ...           ▼                               
+///                                                                                          
+///                                                       merge  ─ ─ ─ ─ ─ ─ ▶ sorted output
+///                                                                               stream     
+///                                                          ▲                               
+///   ...                   ...                ...           │                               
+/// ┌─────┐                ┌─────┐                                                           
+/// │  3  │                │  3  │                           │                               
+/// │  1  │─ ▶  sort  ─ ─ ▶│  1  │─ ─ ─ ─ ─ ─▶ merge  ─ ─ ─ ─                                
+/// └─────┘                └─────┘                                                           
+/// ┌─────┐                ┌─────┐                ▲                                          
+/// │  4  │                │  3  │                                                           
+/// │  3  │─ ▶  sort  ─ ─ ▶│  4  │─ ─ ─ ─ ─ ─ ─ ─ ┘                                          
+/// └─────┘                └─────┘                                                           
+///                                                                                          
+/// in_mem_batches                   do a series of merges that                              
+///                                  each has a limited fan-in                               
+///                                  (number of inputs)                                      
+/// ```
+///
+/// The cascade is built using a series of streams, each with a different purpose:
+///   * Streams leading into the leaf nodes:
+///      1. [`BatchCursorStream`] yields the initial cursors and batches. (e.g. a RowCursorStream)
+///      2. [`BatchTrackingStream`] collects the batches, to avoid passing those around. Yields a [`CursorStream`](super::stream::CursorStream).
+///      3. This initial CursorStream is for a number of partitions (e.g. 100).
+///      4. The initial single CursorStream is shared across multiple leaf nodes, using [`OffsetCursorStream`].
```

Comment:
Not in the notes: reason why a single input CursorStream is shared across leaves, is such that they share the same RowConverter.

Reply:
Mutex removed. This explainer is also updated to reflect the latest design.